src/events/upgrade.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for handling OpensearchDashboards in-place upgrades."""
+import logging
+import time
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    BaseModel,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+)
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+)
+from typing_extensions import override
+
+if TYPE_CHECKING:
+    from charm import (
+        OpensearchDashboardsCharm,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+    )
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchDashboardsDependencyModel(BaseModel):
+    """Model for OpensearchDashboards Operator dependencies."""
+
+    osd_upstream: DependencyModel
+
+
+class ODUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "OpensearchDashboardsCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        pass
```

Comment:
I'm wondering if we want to have a pre-post upgrade check.

In fact, what I would suggest is to leave this empty at the moment, and get back to this in https://warthogs.atlassian.net/browse/DPE-4226, where we are planning to implement a service healthcheck. I think that's the thing that should be executed before and after an upgrade.

Reply:
Regarding Opensearch upgrades:
We are putting a requirement on the Dashboards Application version (which is linked to an Opensearch version).
However that's indeed not a dynamic cross-service check.

Question #1: Do we want to add this feature (here)? I recall we both had a taks on aligning Opensearch and Dashboards versions (but I'm struggling to find it in Jira now). At the time we decided not to get that task done.

Question #2: According to [Charm Interfaces](https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/opensearch_client/v0), currently Opensearch is not publishing vesion information. At a quick check I couldn't find a corresponding API call either (but I may be missing something at that level). How could we check the version or update status of Opensearch?

Regarding "green" state:
Yes, I agree. However the "green" services healthcheck is to be implemented within the scope of the task mentioned in my comment. So I think to re-visit the `pre_upgrade_check` after the dependency task is completed.

Reply:
Summarizing live discussions: we said we would block on upgrading to a higher workload version than what Opensearch may have.

Yes, I also found it in the meanwhile that we communicate the vesion in the relation, and that's what I'm planning to use for the check. 

However a related question (outside of this PR): shouldn't we update [Charm Interfaces](https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/opensearch_client/v0)?

Reply:
@juditnovak approved the PR but I suggest we keep this comment open as we will likely revisit these discussions at `v2`

Reply:
Sure. Thank you.

Reply:
Just to note on this thread: the current mechanism surely works :rofl: 

https://github.com/canonical/opensearch-dashboards-operator/actions/runs/9414634059/job/25939480084

Pinning down Opensearch in the pipelines to the last 2.12 releasse :-)

Reply:
> shouldn't we update [Charm Interfaces](https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/opensearch_client/v0)?

Absolutely! We should add the `kibanaserver` role there

Reply:
...together with the vresion number being exposed.

I craete a ticket so we dont' foget.

Reply:
https://warthogs.atlassian.net/browse/DPE-4576